### ISSUE TITLE
[Bug]  Searching and replacing something in all files once for all also affects high-tech PC and devours 15GB RAM

### ISSUE THREAD

**Issue by gusbemacbe** [microsoft/vscode#89300]
Hello, and @sandy081 and @roblourens!

There is already the former issue the whgich I posted, but your bot closed and locked. Here is https://github.com/microsoft/vscode/issues/74669. 

- VSCode Version: Insiders 1.42.0
- OS Version: Arch Linux Linux Kernel 5.4.12

The steps are same as https://github.com/microsoft/vscode/issues/74669. The difference is that my new PC is a super PC, very powerful and has 16GB RAM memory. When I applied these steps, it devoured 15GB RAM memory and slowed the whole computer down. 

| Piece | Name |
|:--:| :--:|
| DE        | KDE Plasma 5.17.5                 |
| CPU       | AMD Ryzen 7 2700X (16) @ 3.700GHz |
| GPU       | NVIDIA GeForce GTX 1660           |
| RAM       | 16GB DDR4                         |
| Kernel    | 5.4.12-arch1-1-ARCH x86_64        |

**Comment by roblourens** [comment]
I still can't reproduce any issue. It doesn't really make sense to me that it would use 15gb of ram while doing this because these files are not large and 1200 isn't that many.

Can you try this? Do the same steps with a query that matches fewer files. Try to simplify the case, and when you have a case where it doesn't hang, then open the developer tools (Help > Toggle Developer Tools) and use the performance tab to get a profile of the renderer process. This would help us figure out whether there is something very strange happening on your machine. Thanks!

**Comment by gusbemacbe** [comment]
Hello @roblourens !

Sorry for my long absence. My CEO and my CTO were too abusive and authoritative, and did not allow me to answer and help anyone, and overcharged me with a lot, a lot and a lot of things for them. 

I recorded a video, and saved the profiles. I compressed them and provide it for you. If you are not satisfied with what you needed, I can try to record again, enabling memory and save the profiles. 

[Result.zip](https://github.com/microsoft/vscode/files/4184230/Result.zip)

**Comment by gusbemacbe** [comment]
Discard the Result.zip and Update: [Result 2.zip](https://github.com/microsoft/vscode/files/4184247/Result.2.zip) (I forgot of adding last two profiles)

**Comment by roblourens** [comment]
Thanks for doing that. The profile is a little difficult to interpret. Can you tell me what steps you did for the 4 profiles? And the screen recording starts 25 seconds in, what did you do before starting the recording? Had you already clicked "replace all" at that point?

**Comment by gusbemacbe** [comment]
> Can you tell me what steps you did for the 4 profiles? 

The same steps in the #74669:

- Press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>F</kbd>;
- Put `^(\s)*$\n` in the search box and enable regular expression;
- Leave empty replace box and select to replace all once for all.

> And the screen recording starts 25 seconds in, what did you do before starting the recording?
 I started to record the performance and waited, then I clicked to replace all before starting to record the video

> Had you already clicked "replace all" at that point?
Yes

If you need me to record the video before the beginning, here is another video with another single profile renamed as Profile #1:

[VSCode - 2020-02-12.zip](https://github.com/microsoft/vscode/files/4191580/VSCode.-.2020-02-12.zip)

From 50 seconds, it will become increasingly slower, we had to be patient and wait for more than 50 seconds.

**Comment by roblourens** [comment]
Thanks. I don't think I've asked this yet - could you try launching vscode with `--disable-extensions`, and check whether it's any faster without extensions? It seems like extensions might be participating in the 1200 file opens and saves.

**Comment by gusbemacbe** [comment]
Yes, I have just done. Indeed, it is fast, it took just less than 40 seconds in comparison to one or two minutes of previous tests. Here is:

[VSCode 3 – Without extensions.zip](https://github.com/microsoft/vscode/files/4193791/VSCode.3.Without.extensions.zip)

**Comment by roblourens** [comment]
Could you share your extensions list? You can get it by going to Help > Report Issue and starting to report another issue (but just copy the list into this issue)

**Comment by gusbemacbe** [comment]
Here is:

<details>
  <summary>System informations</summary>

| CPUs            | AMD Ryzen 7 2700X Eight-Core Processor (16 x 3372)                                                                                                                                                                                                                                                                                                                                                                             |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| GPU Status      | 2d_canvas: enabledflash_3d: enabledflash_stage3d: enabledflash_stage3d_baseline: enabledgpu_compositing: enabledmultiple_raster_threads: enabled_onoop_rasterization: disabled_offprotected_video_decode: unavailable_offrasterization: disabled_softwareskia_renderer: disabled_off_okvideo_decode: unavailable_offviz_display_compositor: enabled_onviz_hit_test_surface_layer: disabled_off_okwebgl: enabledwebgl2: enabled |
| Load (avg)      | 1, 1, 1                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Memory (System) | 15.65GB (5.56GB free)                                                                                                                                                                                                                                                                                                                                                                                                          |
| Process Argv    |
| Screen Reader   | no                                                                                                                                                                                                                                                                                                                                                                                                                             |
| VM              | 0%                                                                                                                                                                                                                                                                                                                                                                                                                          |

</details>

<details>
  <summary>Extensions</summary>


| Extension                      | Author (truncated) | Version      |
| ------------------------------ | ------------------ | ------------ |
| apacheconf-snippets            | eim                | 1.3.0        |
| atom-keybindings               | ms-                | 3.0.6        |
| autodocstring                  | njp                | 0.4.0        |
| banner-comments-plus           | lun                | 0.3.1        |
| beautify                       | Hoo                | 1.5.0        |
| bracket-pair-colorizer         | Coe                | 1.0.61       |
| calculate                      | aca                | 2.1.0        |
| change-case                    | wma                | 1.0.0        |
| change-color                   | Dee                | 0.0.1        |
| chat                           | kar                | 0.31.0       |
| code-runner                    | for                | 0.9.15       |
| color-highlight                | nau                | 2.3.0        |
| color-tabs-vscode-ext          | ore                | 0.10.0       |
| colored-regions                | mih                | 0.0.3        |
| cpptools                       | ms-                | 0.26.3       |
| csharp                         | ms-                | 1.21.11      |
| css-grid-snippets              | oha                | 0.0.5        |
| css-responsive                 | mre                | 1.1.1        |
| csv-to-table                   | php                | 1.2.3        |
| debugger-for-chrome            | msj                | 4.12.6       |
| emmet-live                     | yse                | 1.0.0        |
| emoji-code                     | idl                | 0.8.2        |
| file-properties-viewer         | bru                | 0.3.4        |
| firebase-explorer              | jsa                | 0.3.3        |
| firebase-snippets              | has                | 0.0.1        |
| ginfuru-vscode-jekyll-syntax   | gin                | 0.1.0        |
| githistory                     | don                | 0.5.0        |
| github-markdown-preview        | bie                | 0.0.2        |
| gitlab-workflow                | fat                | 2.2.0        |
| Go                             | ms-                | 0.13.0       |
| gruvbox-material               | sai                | 5.0.4        |
| haml                           | vay                | 1.0.1        |
| highlight-matching-tag         | vin                | 0.9.6        |
| html-css-class-completion      | Zig                | 1.19.0       |
| html-snippets                  | abu                | 0.2.1        |
| java                           | red                | 0.55.1       |
| jshint                         | dba                | 0.10.21      |
| kotlin                         | fwc                | 0.2.11       |
| Kotlin                         | mat                | 1.7.1        |
| l13-diff                       | L13                | 0.21.0       |
| latex-workshop                 | Jam                | 8.7.2        |
| LiveServer                     | rit                | 5.6.1        |
| markdown-all-in-one            | yzh                | 2.7.0        |
| markdown-checkbox              | bie                | 0.1.3        |
| markdown-extended              | jeb                | 1.0.16       |
| markdown-kbd                   | jja                | 0.2.0        |
| markdown-preview-github-styles | bie                | 0.1.6        |
| markdown-shortcuts             | mdi                | 0.12.0       |
| markdown-table-formatter       | fcr                | 2.0.3        |
| markdown-yaml-preamble         | bie                | 0.0.4        |
| markdownfromcsv                | joj                | 1.0.3        |
| material-icon-theme            | PKi                | 3.9.3        |
| minify                         | Hoo                | 0.4.3        |
| name-that-color                | gui                | 0.1.2        |
| npm-intellisense               | chr                | 1.3.0        |
| php-cs-fixer                   | jun                | 0.1.129      |
| php-debug                      | fel                | 1.13.0       |
| php-intellisense               | fel                | 2.3.14       |
| phpserver                      | bra                | 2.4.6        |
| python                         | ms-                | 2020.2.62710 |
| rainbow-csv                    | mec                | 1.5.0        |
| spellright                     | ban                | 3.0.50       |
| svg                            | joc                | 1.1.1        |
| swdc-vscode                    | sof                | 2.0.9        |
| theme-by-language              | jsa                | 1.1.1        |
| vsc-material-theme             | Equ                | 32.4.0       |
| vsc-material-theme-icons       | equ                | 1.1.3        |
| vscode-apache                  | mrm                | 1.2.0        |
| vscode-attrs-sorter            | mrm                | 2.1.0        |
| vscode-colorize                | kam                | 0.8.15       |
| vscode-css-peek                | pra                | 3.0.2        |
| vscode-csscomb                 | mrm                | 5.3.0        |
| vscode-custom-css              | be5                | 3.0.9        |
| vscode-duplicate               | mrm                | 1.2.1        |
| vscode-edit-csv                | jan                | 0.2.5        |
| vscode-eslint                  | dba                | 2.0.15       |
| vscode-firefox-debug           | fir                | 2.6.1        |
| vscode-google-translate        | fun                | 1.4.2        |
| vscode-jekyll-snippets         | gin                | 0.9.1        |
| vscode-language-pack-en-GB     | MS-                | 1.42.2       |
| vscode-mjml                    | att                | 1.6.0        |
| vscode-npm-script              | eg2                | 0.3.11       |
| vscode-tailwindcss             | bra                | 0.2.0        |
| vscode-tldr                    | bmu                | 1.0.0        |
| vscode-wakatime                | Wak                | 2.3.0        |
| vscode-xml                     | red                | 0.10.1       |
| vsfire                         | tob                | 1.4.1        |
| vsliveshare                    | ms-                | 1.0.1510     |
| vsliveshare-audio              | ms-                | 0.1.80       |

</details>

<details>
  <summary>Running processes</summary>

```
CPU %	Mem MB	   PID	Process
    0	   128	192067	code-insiders main
    0	    32	192071	   zygote
    0	     0	192073	     zygote
    0	   160	192097	   gpu-process
    0	    16	192139	     broker
    0	    64	192104	   utility
    0	   256	192126	   window (16 - Visual Studio Code - Insiders)
    0	   305	193182	     extensionHost
    0	    48	194927	       /opt/visual-studio-code-insiders/code-insiders /opt/visual-studio-code-insiders/resources/app/extensions/json-language-features/server/dist/jsonServerMain --node-ipc --clientProcessId=193182
    0	    80	193243	     watcherService
    0	    48	193263	     searchService
    0	   112	193213	   shared-process
    0	     0	197457	     /usr/bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
    0	    96	197378	   window (Issue Reporter)
```

</details>

<details>
  <summary>Workspace</summary>

```
|  Window (16 - Visual Studio Code - Insiders)
|    Folder (16): 1341 files
|      File types: svg(1341)
|      Conf files:
```

</details>

**Comment by roblourens** [comment]
That's a lot :)

I will have to install some of these and test them out when I have some time.

**Comment by gusbemacbe** [comment]
@roblourens, the solution is to create each workspace with few specific extensions for specific project, to keep the performance up. But our goal is to run your app with how many extension we have without slowing down. I know it is an utopia.

**Comment by roblourens** [comment]
I agree, it's just that one of these extensions is probably doing something slow that it shouldn't be doing. I'll try to figure out which one it is when I get a chance but if you want to try disabling extensions to narrow down the list, that would be very helpful too.

**Comment by roblourens** [comment]
@sandy081  is there some way we can do the replace without publicly registering text models so that extensions don't find out about it, for example I see that `onDidOpenTextDocument` is fired and maybe extensions do some work based on that.

Another result is that I get all the file extension-based marketplace extension tips when I do a replace, for all the file extensions that are touched by the replace, which could be a little confusing.

**Comment by sandy081** [comment]
@roblourens Not much insights on it. I use the internal bulk edit api. 
@jrieken is it possible?

**Comment by jrieken** [comment]
>  I use the internal bulk edit api. @jrieken is it possible?

Sure, not use that API 🤓 But then you won't get undo and all these things

**Comment by sandy081** [comment]
@roblourens Don't want to reinvent different bulk edit. So its not possible.

**Comment by jrieken** [comment]
You can add a cap on the number of replaces that you support

**Comment by roblourens** [comment]
This can happen at a pretty low number of replaces.

**Comment by jvilk-stripe** [comment]
I opened up https://github.com/microsoft/vscode/issues/101523 (closed as dupe) because this is a perf problem even with extensions disabled (but is exacerbated when you have linting extensions).

It's also surprising to me as an extension developer; I expect files to be opened at a human rate. Some of our internal extensions don't react well to 1000 files being opened at once, and we're tweaking them to handle this use case.

**Comment by roblourens** [comment]
Merging this into https://github.com/microsoft/vscode/issues/64485
